### PR TITLE
gh-91731: Don't define 'static_assert' in C++11 where is a keyword to avoid UB

### PR DESCRIPTION
**Pull request by pablogsal**

### CODE REVIEW

File: Include/pymacro.h [hunk 1]
```diff
@@ -14,7 +14,7 @@
 // compiler support (gcc >= 4.6) and is called _Static_assert.
 #if (defined(__GLIBC__) \
      && (__GLIBC__ < 2 || (__GLIBC__ == 2 && __GLIBC_MINOR__ <= 16)) \
```
Comment (AA-Turner) [inline Include/pymacro.h:17]: Ordering suggestion -- put all version checks first (had to split the suggestion due to the GH UI). I'm not sure what precedent here is though.
```suggestion
     && (__GLIBC__ < 2 || (__GLIBC__ == 2 && __GLIBC_MINOR__ <= 16)) \
     && !(defined(__cplusplus) && __cplusplus >= 201103L) \
```

File: Include/pymacro.h [hunk 2]
```diff
@@ -14,7 +14,7 @@
...
-     && !defined(static_assert))
+     && !defined(static_assert) && !(defined(__cplusplus) && __cplusplus >= 201103L))
```
Comment (AA-Turner) [inline Include/pymacro.h]:
```suggestion
     && !defined(static_assert))
```

File: Include/pymacro.h [hunk 3]
```diff
@@ -14,7 +14,7 @@
 // compiler support (gcc >= 4.6) and is called _Static_assert.
```
Comment (AA-Turner) [inline Include/pymacro.h:14]:
```suggestion
// compiler support (gcc >= 4.6) and is called _Static_assert.
// In C++ 11 static_assert is a keyword, redefining is undefined behaviour.
```

File: Misc/NEWS.d/next/Core and Builtins/2022-06-10-23-41-48.gh-issue-91731.fhYUQG.rst [hunk 4]
```diff
@@ -0,0 +1,3 @@
+Avoid defining the ``static_assert`` when compiling with C++ 11, where this
```
Comment (vstinner) [inline Misc/NEWS.d/next/Core and Builtins/2022-06-10-23-41-48.gh-issue-91731.fhYUQG.rst]: You can move this NEWS entry to Build, or C API, category.

**Review by vstinner — APPROVED** [review]
LGTM.

**Pull request merged**